src/Core/MySQL/MySQLJsonConv.cpp

```diff
@@ -0,0 +1,940 @@
+ * offset). When using the large storage format, it is 5 (1 byte for type, 4
+ * bytes for offset).
+ */
+constexpr UInt8 VALUE_ENTRY_SIZE_SMALL = 1 + SMALL_OFFSET_SIZE;
+constexpr UInt8 VALUE_ENTRY_SIZE_LARGE = 1 + LARGE_OFFSET_SIZE;
+
+static void checkLength(size_t len, size_t wanted)
+{
+    if (len < wanted)
+    {
+        throw DB::Exception(ErrorCodes::UNKNOWN_EXCEPTION, "illegal json data");
+    }
+}
+
+
+static UInt8 getUInt8(const char * buffer)
+{
+    UInt8 ret;
+    memcpy(&ret, buffer, sizeof(ret));
+    return ret;
+}
+
+static Int16 getInt16(const char * buffer)
+{
+    Int16 ret;
+    memcpy(&ret, buffer, sizeof(ret));
+    return ret;
+}
+
+static UInt16 getUInt16(const char * buffer)
+{
+    UInt16 ret;
+    memcpy(&ret, buffer, sizeof(ret));
+    return ret;
+}
+
+static Int32 getInt32(const char * buffer)
+{
+    Int32 ret;
+    memcpy(&ret, buffer, sizeof(ret));
+    return ret;
+}
+
+static UInt32 getUInt32(const char * buffer)
+{
+    UInt32 ret;
+    memcpy(&ret, buffer, sizeof(ret));
+    return ret;
+}
+
+static Int64 getInt64(const char * buffer)
+{
+    Int64 ret;
+    memcpy(&ret, buffer, sizeof(ret));
+    return ret;
+}
+
+static UInt64 getUInt64(const char * buffer)
+{
+    UInt64 ret;
+    memcpy(&ret, buffer, sizeof(ret));
+    return ret;
+}
+
+static double getDouble(const char * buffer)
+{
+    double ret;
+    memcpy(&ret, buffer, sizeof(ret));
+    return ret;
+}
+
+static bool readVariableLength(const char *data, size_t data_length, UInt32 *length, UInt8 *num)
+{
+    /*
+    It takes five bytes to represent UINT_MAX, which is the largest
+    supported length, so don't look any further.
+    */
+    const size_t max_bytes = std::min(data_length, static_cast<size_t>(5));
+
+    size_t len = 0;
+
+    for (size_t i = 0; i < max_bytes; i++)
+    {
+        // Get the next 7 bits of the length.
+        len |= (data[i] & 0x7f) << (7 * i);
+        if ((data[i] & 0x80) == 0)
+        {
+            // The length shouldn't exceed 32 bits.
+            if (len > UINT_MAX)
+                return true; /* purecov: inspected */
+
+            // This was the last byte. Return successfully.
+            *num = static_cast<UInt8>(i + 1);
+            *length = static_cast<UInt32>(len);
+            return false;
+        }
+    }
+
+    // No more available bytes. Return true to signal error.
+    return true;
+}
+
+static size_t readOffsetOrSize(const char * buffer, bool large)
+{
+    return large ? getUInt32(buffer) : getUInt16(buffer);
+}
+
+
+static JsonValue parseArrayOrObject(JsonEnumType json_type, const char * data, size_t len, bool large)
+{
+    size_t offset_size = large ? LARGE_OFFSET_SIZE : SMALL_OFFSET_SIZE;
+
+    checkLength(len , 2 * offset_size);
+    auto element_count = readOffsetOrSize(data, large);
+
+    auto bytes = readOffsetOrSize(data + offset_size, large);
+
+    if (bytes > len)
+    {
+        throw DB::Exception(ErrorCodes::UNKNOWN_EXCEPTION, "illegal json data");
```

Comment:
I see that the error messages are all the same, but this is wrong - we need to distinguish the error cases.